### PR TITLE
Support array indexes

### LINKED ISSUE #71: Support array index access
I ran into this issue when upgrading from 4->6, which of course is a breaking change, but I _think_ the new behavior is inconsistent. 🙇 

Consider some objects that have arrays, like:

```js
const emptyArray = { foo: [] }
const oneItem = { foo: [ { fizz: 'buzz' } ] }
```

I want to access properties on the array's objects, and provide a default value if it doesn't exist.

```js
// passes: default value is found
t.is(dotProp.get(emptyArray, 'foo.0.fizz', 'bazz'), 'bazz')
// passes: existing value is found
t.is(dotProp.get(oneItem, 'foo.0.fizz', 'bazz'), 'buzz')
```

However, in 4->6 the behavior changed when accessing an array index that doesn't exist:

```js
// passes in 4 but fails in 6
t.is(dotProp.get({ foo: [ 'bar' ] }, 'foo.1', 'bazz'), 'bazz')
// passes in 6 but fails in 4
t.is(dotProp.get({ foo: [ 'bar' ] }, 'foo.1', 'bazz'), undefined)
```

I think that this behavior is ... inconsistent, probably?

If accessing a deeper property in an array element triggers a default value fallback, shouldn't accessing an array element trigger a fallback if that index doesn't exist?

### LINKED ISSUE #87: How to access data with array
Trying to do something like this:

```javascript
const dotProp = require('dot-prop')

const data = {
  data: [{ price: 1234 }],
}

const result = dotProp.get(data, 'data[0].price')
```

But result is always `undefined`

Is this possible with this library? Or perhaps I should look at an alternative?

### PR DESCRIPTION
**Pull request by Richienb**
> Waiting for #80 before updating typings

# Breaking change

Before:

```js
dotProp.set({}, 'a.0', true);
//=> {a: {'0': true}}
```

After:

```js
dotProp.set({}, 'a[0]', true);
//=> {a: [true]}
```

Fixes #71 
Fixes #87

### CODE REVIEW

**Comment by sindresorhus** [comment]
I think we should use `a[0]` as it's a more common syntax.

**Comment by Richienb** [comment]
What should `a.[0].b` instead of `a[0].b` give?

**Comment by sindresorhus** [comment]
> What should a.[0].b instead of a[0].b give?

IMHO, nothing. It would be the same as `object['[0]']`. We could maybe throw an error to let the user know they made a mistake.

**Comment by Zehir** [comment]
I think that a.0.b make more sense.
If `a` is an array he should return `a[0].b`.
If `a` is an object he should return `a[0].b`because you can't write `a.0.b`.
But for setting new values inside a non existing patch you can't detect if should be an array or an object.
And how you add an element inside a not existing array ? Like `a[].b` when a is currently undefined.
I don't think a lot of users are using `[]` as object keys but yes it's could be a breaking change.
And what about escaping the numbers like `a.\\0.b`. Or using a `..` notation to say that the next value should be in an array.
An other question is if you set a value like `a.5.b` what you put inside the first 5 element of the array ? Undefined ?

To do this without breaking changes you could just not let the user create an new array with the dot notation but let them edit them, add items or get them with classic `a.0.b` notations. And if there is no number just add it to the array.
```javascript
let data = {a:[]}
dotProp.set(data , 'a.0', 1);
console.log(data) // {a:[1]}
dotProp.set(data , 'a.1.b', 2);
console.log(data) // {a:[1, {b:2}]}
dotProp.set(data , 'a..b', 3);
console.log(data) // {a:[1, {b:2}, {b:3}]}
```
```javascript
let data = {}
dotProp.set(data , 'a.0', 1); // Throw exception
```
Or with the `..` notation:
```javascript
let data = {}
dotProp.set(data , 'a..0', 1);
console.log(data) // {a:[1]}
dotProp.set(data , 'a..1.b', 2);
console.log(data) // {a:[1, {b:2}]}
data = {}
dotProp.set(data , 'a..1.b', 2);
console.log(data) // {a:[undefined, {b:2}]}
```
Currently the `..` add an empty key but that make sence because there is no key between the two dots. So it's again a breaking change..
```javascript
let data = {}
dotProp.set(data , 'a..b', 2);
{ a: { '': { b: 2 } } }
```

**Comment by Richienb** [comment]
@sindresorhus I've managed to turn the segmenting logic into a simple regex. Should I add it directly to the `index.js` file or does its scope warrant a seperate package?

**Comment by sindresorhus** [comment]
@Richienb Are you sure a regex is a good idea for this? I'm generally trying to avoid regexes unless absolutely necessary as it's so easy to accidentally make a ReDoS.

**Comment by Richienb** [comment]
Good point. Regexes make it much easier to implement this extra feature but otherwise, we could just revert to scanning the string from left to right to segment it.

**Comment by sindresorhus** [comment]
Bump, in case you forgot about this. If you're just busy, feel free to ignore.

**Comment by Richienb** [comment]
@sindresorhus Are you ok with the parsing strategy I mentioned above?

**Comment by sindresorhus** [comment]
> we could just revert to scanning the string from left to right to segment it.

👍

File: index.js [hunk 1]
```diff
@@ -7,25 +7,136 @@ const disallowedKeys = new Set([
...
+						parts.push(currentPathSegment);
+					} else {
+						parts.push(index);
+					}
+
+					currentPathSegment = '';
+					break;
+				}
+
+				// Falls through
```
Comment (Richienb) [inline index.js:114]: https://eslint.org/docs/rules/no-fallthrough

File: index.js [hunk 2]
```diff
@@ -43,12 +154,20 @@ module.exports = {
 		}
 
 		for (let i = 0; i < pathArray.length; i++) {
-			object = object[pathArray[i]];
+			const key = pathArray[i];
+			const index = Number.parseInt(key, 10);
+
+			// Disallow string indexes
+			if (!Number.isInteger(key) && Array.isArray(object) && !Number.isNaN(index) && object[index] === object[key]) {
+				object = i === pathArray.length - 1 ? undefined : null;
```
Comment (Richienb) [inline index.js:162]: `null` is used since the rest of the code already recognises what it means

File: index.js [hunk 3]
```diff
@@ -43,12 +154,20 @@ module.exports = {
 		}
 
 		for (let i = 0; i < pathArray.length; i++) {
-			object = object[pathArray[i]];
+			const key = pathArray[i];
+			const index = Number.parseInt(key, 10);
+
+			// Disallow string indexes
+			if (!Number.isInteger(key) && Array.isArray(object) && !Number.isNaN(index) && object[index] === object[key]) {
```
Comment (Richienb) [inline index.js:161]: Because JavaScript automatically coerces `object["1"]` to `object[1]` we need to detect that and override that default behaviour

**Comment by Richienb** [comment]
I have chosen to segment `foo[bar]` as `['foo', 'bar']`.

**Comment by Richienb** [comment]
The tests are failing because of a bug: https://github.com/sindresorhus/dot-prop/pull/86

**Comment by Richienb** [comment]
CI is now passing.

File: test.js [hunk 4]
```diff
@@ -26,6 +26,57 @@ test('get', t => {
 	t.true(dotProp.get({'foo\\.bar': true}, 'foo\\\\.bar'));
 	t.is(dotProp.get({foo: 1}, 'foo.bar'), undefined);
 
+	t.true(dotProp.get([true, false, false], '[0]'));
```
Comment (sindresorhus) [inline test.js:29]: The array access stuff should be in separate tests. This test is already way too large.

File: test.js [hunk 5]
```diff
@@ -26,6 +26,57 @@ test('get', t => {
...
+	t.true(dotProp.get([{foo: [true]}], '[0].foo[0]'));
+	t.true(dotProp.get({foo: [0, {bar: true}]}, 'foo[1].bar'));
+
+	t.false(dotProp.get(['a', 'b', 'c'], '3', false));
+	t.false(dotProp.get([{foo: [1]}], '[0].bar[0]', false));
+	t.false(dotProp.get([{foo: [1]}], '[0].foo[1]', false));
+	t.false(dotProp.get({foo: [0, {bar: 2}]}, 'foo[0].bar', false));
+	t.false(dotProp.get({foo: [0, {bar: 2}]}, 'foo[2].bar', false));
+	t.false(dotProp.get({foo: [0, {bar: 2}]}, 'foo[1].biz', false));
+	t.false(dotProp.get({foo: [0, {bar: 2}]}, 'bar[0].bar', false));
```
Comment (sindresorhus) [inline test.js:39]: Can you add a bunch of tests that tries to trick the parsing? With escapes in different places and stuff. I'm quite paranoid that this change will introduce some kind of vulnerability.

File: index.d.ts [hunk 6]
```diff
@@ -55,6 +58,10 @@ declare const dotProp: {
 	dotProp.set(object, 'foo.baz', 'x');
 	console.log(object);
 	//=> {foo: {bar: 'b', baz: 'x'}}
+
+	dotProp.set(object, 'foo.biz.0', 'a');
```
Comment (sindresorhus) [inline index.d.ts:62]: This syntax is leftover. The `.0` part.

File: index.js [hunk 7]
```diff
@@ -74,7 +195,11 @@ module.exports = {
 			const p = pathArray[i];
 
 			if (!isObject(object[p])) {
-				object[p] = {};
+				if (Number.isInteger(pathArray[i + 1])) {
+					object[p] = [];
+				} else {
+					object[p] = {};
+				}
```
Comment (sindresorhus) [inline index.js:202]: This would read simpler as a ternary.

File: index.js (hunk 2 above)
Comment (sindresorhus) [reply index.js:164]: Maybe add a code comment. It might be tempting to try to remove the `null` sometime in the future.

File: index.js (hunk 3 above)
Comment (sindresorhus) [reply index.js:163]: This would also be good as a code comment.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
I would like more tests with lots of different variations of invalid patterns (including different kind of escaping), just to ensure the parsing handles all edge-cases.

**Comment by sindresorhus** [comment]
> I have chosen to segment foo[bar] as ['foo', 'bar'].

Why? I think that should be an invalid pattern and not return anything. It's beneficial to keep the syntax as simple as possible to reduce the chances of bugs and vulnerabilities.

**Comment by Richienb** [comment]
> > I have chosen to segment foo[bar] as ['foo', 'bar'].
> 
> Why? I think that should be an invalid pattern and not return anything. It's beneficial to keep the syntax as simple as possible to reduce the chances of bugs and vulnerabilities.

I'm trying to be as lenient as possible on the errors so instead, I'll have this literally parsed as ['foo[bar]']

**Comment by sindresorhus** [comment]
Bump :)

File: test.js [hunk 8]
```diff
@@ -56,6 +61,78 @@ test('get', t => {
 	F4Class.prototype.foo = 1;
 	const f4 = new F4Class();
 	t.is(dotProp.get(f4, 'foo'), 1); // #46
+
+	t.true(dotProp.get({'': {'': true}}, '..'));
+	t.true(dotProp.get({'': true}, '.'));
```
Comment (Richienb) [inline test.js:66]: Not sure what these should actually do

Comment (sindresorhus) [reply test.js:66]: They should just return `undefined`. No point in supporting weird cases like that.

**Comment by sindresorhus** [comment]
Bump => https://github.com/sindresorhus/dot-prop/pull/82#discussion_r766108928

Happy new year :)

**Comment by Richienb** [comment]
I realised that an extremely lenient parser would probably lead to user confusion so I made it much stricter.

File: index.js [hunk 9]
```diff
@@ -7,30 +7,156 @@ const disallowedKeys = new Set([
 	'constructor'
 ]);
 
-const isValidPath = pathSegments => !pathSegments.some(segment => disallowedKeys.has(segment));
+const digits = new Set('0123456789');
```
Comment (Richienb) [inline index.js:10]: Either this or `String#includes`

Comment (sindresorhus) [reply index.js:10]: What you have is better.

**Comment by sindresorhus** [comment]
Great work, @Richienb 👍
